### ISSUE TITLE
App Level Middleware Invocation and Route Ordering

### ISSUE THREAD

**Issue by shhmendez** [mahmoud/clastic#28]
Hey guys, I'm trying to implement a cors middleware that is always called. To my understanding, middleware is only called when a route is matched. So I tried creating a route that matches everything

```
class Cors(Middleware):
    def render(self, next, context, request, _route):
        res = render_basic(context,request, _route)
        res.headers['Access-Control-Allow-Origin'] = "*"
        res.headers['Access-Control-Allow-Headers'] = "*"
        return res

routes = [("/test",lambda: None),("/.*",lambda: None) ]

app = Application(routes,middlewares=[Cors()])
```

The cors middleware only gets called when a route is matched, and because `"/.*"` matches everything `"/test"` can never be reached

So the problem is 2 fold:
1. I need middleware that is called on every server hit OR
2. I need a route that match everything except routes that have already been defined

**Comment by mahmoud** [comment]
This is such a good question. This is absolutely possible and just requires a couple clarification.

1. All clastic applications have a built-in ["null" route](https://github.com/mahmoud/clastic/blob/fa1a3b4e66cbc350ebd2b625f624e541fa6ef81e/clastic/application.py#L191), which is always at the end of the route list, and exists to raise 404s and the like. Because of this, all of a given application's middlewares are always evaluated for every request. 
1. In your middleware, I've noticed you've put all your logic in the render step. Note that this will only execute if a route matches *and* returns a data structure (as opposed to a returning a `Response` or raising an exception). If you want something to always  execute, you should put it on the `.request()`, before calling `next()`. (Also, don't forget to call next in your mw :) )
1. NullRoutes are just normal routes. They match any path because they use a path match pattern, e.g., `/<path*>`. That'll match any URL, so be careful where you put it on your route list. If it goes at the end, it'll be that "a route that match everything except routes that have already been defined" that you want.

So in short, it's totally possible to write a middleware that gets called on every hit, as well as a route that matches everything.

**Comment by shhmendez** [comment]
Awesome! It seems I’m little bit confused about route ordering then.
I was under the impression that the application matched a routes in the order provided in the list. That seems to be the case based on your comment about nullroute 

> If it goes at the end, it'll be that "a route that match everything except routes that have already been defined" that you want.

So why doesn’t my `“/.*”` route at the end of my route list behave like a nullroute ?

**Comment by mahmoud** [comment]
I'm not in a spot where I can test this, but I think that's just a syntax fix. Change `"/.*"` to `"/<path*>"`. The exact name of the variable provided doesn't matter but all parts of clastic url patterns must be named. :)

**Comment by shhmendez** [comment]
That seems to work thanks! I still have questions about a couple different things, and I am keeping `Cors.render` because I don't know of any other way to access the headers in the response. 

For completion the final change was simply the route path
```
class Cors(Middleware):
    def render(self, next, context, request, _route):
        res = render_basic(context,request, _route)
        res.headers['Access-Control-Allow-Origin'] = "*"
        res.headers['Access-Control-Allow-Headers'] = "*"
        return res

routes = [("/test",lambda: None),("/<path*>",lambda: None) ]

app = Application(routes,middlewares=[Cors()])
```

Looking forward to further discussions.

**Comment by mahmoud** [comment]
Here's a quick sketch, might work, might need a tweak or two :)

```python
class Cors(Middleware):
    def render(self, next, context, request, _route):
        resp = render_basic(context,request, _route)
        return resp

    def request(self, next):
        resp = next()
        resp.headers['Access-Control-Allow-Origin'] = "*"
        resp.headers['Access-Control-Allow-Headers'] = "*"
        return resp
```
